blockchain/blockchain.go

```diff
@@ -111,37 +109,30 @@ type (
 	}
 )
 
-// ProductivityByEpoch returns the map of the number of blocks produced per delegate in an epoch
-// TODO: move to poll protocol and implement reading current epoch meta from state factory -- now only reading current epoch productivity
-func ProductivityByEpoch(ctx context.Context, bc Blockchain, sr protocol.StateReader, epochNum uint64) (uint64, map[string]uint64, error) {
+// CurrentEpochProductivity returns the map of the number of blocks produced per delegate in current epoch
+// TODO: implement reading current epoch meta from state factory
+func CurrentEpochProductivity(ctx context.Context, bc Blockchain, sr protocol.StateReader, epochNum uint64) (uint64, map[string]uint64, error) {
```

Comment:
we shouldn't introduce "sr" in this function.
blockchain.ProductivityByEpoch is something relying on the data stored in blockchain db. If we have all blocks, we should be able to return the productivity for any epoch.

Reply:
According to  #1943 , we are going to read productivity info(producers) from state DB, is it not true?